Initial prototype for pushing and pulling data in to a database file

peterjamesnugent opened the pull request:
<!-- PLEASE ENSURE YOU REVIEW THE CONTENT OF EACH PR CAREFULLY, INCLUDING SUBSEQUENT COMMENTS BY YOURSELF OR OTHERS. -->
<!-- IN PARTICULAR PLEASE ENSURE THAT SENSITIVE OR INAPPROPRIATE INFORMATION IS NOT UPLOADED -->
   
### Issues addressed by this PR
<!-- Add reference(s) to issue(s) solved by this PR. Please use keyword Fixes/Closes as per https://help.github.com/articles/closing-issues-using-keywords/ -->

Closes #1 

<!-- Add short description of what has been fixed -->

### Test files
<!-- Link to test files to validate the proposed changes -->
[Test script](https://burohappold.sharepoint.com/:u:/s/BHoM/EQWVXv8NedNAufop2ZJyvEkBX6fgRT6f0Qgw3NwgSrjMig?e=snHiFl)

### Changelog
<!-- Text to go into changelog if applicable -->
<!-- Please see https://github.com/BHoM/documentation/wiki/changelog for guidelines -->
- Added an `SQLiteAdapter` that allows users to connect to a database file or use an in memory database;
- The `DatabaseMode` is controlled by the `SQLiteSettings` and includes other options such as `EnableWalMode` and `ConnectionTimeOut`;
- Added push and pull functionality for `IObject`s - note that `IObject`s must have primitive properties or use the `PropertyMapping` in the `PushConfig` when pushing to a .db file;
- `PushConfig` also supports mapping of `Fragments` such as `AdapterId` objects;
- `Pull` is supported by `EqualityFilterRequest` and `RangeFilterRequest`, this includes options for multiple sort conditions using `SortColumn`;
- These requests support multiple filters per request and users can dictate logic;
- `Open` and `Close` `Execute` actions are implemented;
- A set of `Commands` are include in `Engine.Compute` to generate sql commands that can be used with the `Execute` methods;
- `Remove` functionality supported using filters mentioned above to remove entries from the database.

### Additional comments
<!-- As required -->

peterjamesnugent committed:
Update .gitignore and .csproj files

peterjamesnugent committed:
Add enums for different settings in the SQLite_Adapter

peterjamesnugent committed:
Add SQLiteSettings and configs for tables and queries

peterjamesnugent committed:
Add Requests for querying the adapter.

peterjamesnugent committed:
Add different objects for tables. columns and indexes

peterjamesnugent committed:
Remove example methods, update execute and pull methods - testing check point

peterjamesnugent committed:
Add some testing to verify basic functionality

peterjamesnugent committed:
Fix DLLs post build events for NuGet packages

peterjamesnugent committed:
Split out functionality for TableRequest and SchemaRequest

peterjamesnugent committed:
Add objects for creating table schemas, columns and tables as well as push functionality

peterjamesnugent committed:
Update .gitignore

peterjamesnugent committed:
Add some basic tests

peterjamesnugent committed:
Reorder some parameters

peterjamesnugent committed:
Remove preliminary work for pushing nested BHoMObjects as they are no longer needed for the prototype

peterjamesnugent committed:
Add IRecord (temp),, and methods to assist in getting property paths, determining whether primitive etc.

peterjamesnugent committed:
Add new filter methods and methods to manage them in to SQL queries, add methods for type table and querying it

peterjamesnugent committed:
Create GetTypeName.cs

peterjamesnugent committed:
Add type table creation and initialisation and related methods. Update pull and dealing with FilterRequests

peterjamesnugent committed:
Fix bug where table created was showing multiple times, added schema table popualtion, checks for type management and added checks for primitive data (temporary)

peterjamesnugent committed:
Minor changes to Adapter behaviour and formatting

peterjamesnugent committed:
Rename TableData object.

peterjamesnugent committed:
Split out the ValidatePropertyMappings in to seperate methods

peterjamesnugent committed:
Fixed language mistakes, update push to support Table and TableSchema, combine SQL command in to single method for BHoMObject and SQLIte_oM objects

peterjamesnugent committed:
Make sure schema table is populated, update EqualityFilterRequest

peterjamesnugent committed:
Renove constructors for `GeneralDomain` and fix BHoM_Guid creation

peterjamesnugent committed:
Use a tolerance for comparing numeric values (1E-9) due to precision issues with floats

peterjamesnugent committed:
Remove redundant BHoM_Guid column creation and simplify TableExists method

peterjamesnugent committed:
Add unit tests for added functionality - needs refinement

peterjamesnugent committed:
Store type for conversion back from SQL (this is an issue with bool) and fix bug where excluded properties were not being excluded in IRecord and fallback method

peterjamesnugent committed:
Fix bug where columns were not being read because they contained the word Create and update security logic

peterjamesnugent committed:
Update ValidateColumnName.cs

peterjamesnugent committed:
Remove example object and run tests with a Bar object for the complex object

peterjamesnugent committed:
Remove checks for deeply nested properties of interface objects and defunct tests

peterjamesnugent committed:
Add type check for eNums that are stored as  long, update test assets and use consistent naming for table when overriden by PushConfig

peterjamesnugent committed:
Rename methods, split out in to seperate .cs files and fix descriptions

peterjamesnugent committed:
Remove duplicate methods and consolidate methods

peterjamesnugent committed:
Rename tests

peterjamesnugent committed:
Rename files from American english

peterjamesnugent committed:
Remove defunct methods, consolidate methods under single .cs file and general clean up

peterjamesnugent commented:
@BHoMBot check compliance

peterjamesnugent committed:
Remove unused TypeRegsitration

peterjamesnugent commented:
@BHoMBot check core

peterjamesnugent committed:
Add multiple range filters and update some compliance filters

peterjamesnugent committed:
Move a number of methods to the Adapter so they are not exposed and the Engine does not rely on the SQLite packages

peterjamesnugent commented:
@BHoMBot check compliance

peterjamesnugent committed:
Moving of methods to adapter, changing to private and removal of unnecessary objects

peterjamesnugent committed:
Clear up tests

peterjamesnugent commented:
@BHoMBot check compliance

peterjamesnugent committed:
Fix compliance failures

peterjamesnugent commented:
@BHoMBot check compliance

peterjamesnugent commented:
@BHoMBot check compliance

peterjamesnugent committed:
Compliance fixes

peterjamesnugent commented:
@BHoMBot check compliance

peterjamesnugent commented:
@BHoMBot check installer

peterjamesnugent committed:
Remove duplicate method and fix .csproj file

peterjamesnugent committed:
Rename ColumnSchema method

peterjamesnugent committed:
Add execute command for custom sql commands not supported with existing push/pull and update IsValid method

adecler reviewed (commented):
I'm generally happy with the code and they user workflow in the UI. 
Here's my notes going through the code: 

 - Engine.Compute.Table -> it feel like this could use a better name since it creates a command string and not a Table

 - Why are we limiting the Push to IBHoMObjects. Could it be IObject instead ? Same thing for IRecord.

 - Query.IsPrimitive -> You don't have to do all the work, Type.IsPrimitive is already defined in C#

 - Compute.ValidatePropertyMapping -> The name feels wrong as it implies that a Boolean is expected as return value. 

 - Compute.ResolveColumnSchema -> The separation in 3 tiers don't make sense to me and generate al lot of duplicate code. You can just collapse them into one: get all primitives that are not excluded, then add properties mapped by the user. With that in mind, I don't see the benefit of IRecord.

 - What's the point of the schema table if it is not used when pushing objects ?

 - Query.GetPropertyValueFromPath() -> Why not use the GetPropertyValue method in the core engine ?

 - I'm not a big fan of the Example classes being defined in the oM. They should be in the Tests project instead

 - It is better to make your fallback method for ExecuteCommand private as it should never be called directly. So, making it private avoid the confusing with IExecuteCommand

 - I don't see any way to remove entries from a table. I see two options: 
   - The user has an option (in the push config) to replace existing content with the one they are pushing
   - The user can explicitly delete the content of a table as an Execute command
I see you have a Compute.Delete method but I don't see how this fits into the current workflow

 - A bit strange that EqualityFilterRequest is using a list of ColumnFilter while RangeFilterRequest is using a list of dictionaries of GeneralDomains. Wouldn't it make more sense if they were structured the same way ?

 - The methods EqualityFilterRequest and RangeFilterRequest are very similar (only one line different) so they should probably be handled directly in the Pull method. 

 - FilterResult is a strange name in this case. It is not the result of the FilterRequest as the name would implies. I am also not sure about the fact that is is created in a Convert method. Convert methods are usually used to convert BHoM object to and from external classes.

Happy to discuss further in a call if any of the above needs clarification.

peterjamesnugent committed:
Add support for retrieving fragment information and specifically ids

peterjamesnugent committed:
Update Compute.Table to Compute.TableCommand

peterjamesnugent committed:
Update IsPrimitive() to use the System.IsPrimitive() method first and then catch other cases that are not covered (e.g. string, DateTime etc.) but suitable for SQLite

peterjamesnugent committed:
Rename ValidatePropertyMappings to just PropertyMappings (similar for FragmentMappings)

peterjamesnugent committed:
Update `Execute` fallback method to be `private`. Moved `Example` to test project.

peterjamesnugent committed:
Add `ISqlRequest` to unify the filter requests and simplify the `Pull` method. Update `FilterRequest` to `FilterCommand` to align with wider naming.

peterjamesnugent committed:
Remove `IRecord` and associated methods, revert `Compute.ColumnSchema` to just use Tier 2 strategy. and update tests.

peterjamesnugent committed:
Use `IObject` instead of `BHoMObject` for base class

peterjamesnugent committed:
Add `IDelete` override, unify ExecuteQuery to cover SELECT, DELETE, COUNT, add DeleteConfig object and add tests for Delete

peterjamesnugent committed:
Change EqualityFilter where no values means pull all rows and fix delete tests

peterjamesnugent commented:
Thanks for the comments @adecler, I have implemented most of your suggestions and added some commentary below.

Test script has been updated to cover:
- Push of Node with simple nested mapping;
- Push of Node with fragment mapping;
- Pull using single and multiple equality filter;
- Pull using single and multiple range filter;
- Execute actions for open and close;
- Delete using single equality filter;
- Delete using multiple range filter.

 - Engine.Compute.Table -> it feel like this could use a better name since it creates a command string and not a Table
**Updated to `Engine.Compute.TableCommand` to align with the `Execute.Command` - did think about statement or including sql but command feels cleaner.**

 - Why are we limiting the Push to IBHoMObjects. Could it be IObject instead ? Same thing for IRecord.
**Updated to use `IObject`.**

 - Query.IsPrimitive -> You don't have to do all the work, Type.IsPrimitive is already defined in C#
**It returns `false` for `string`, `DateTime` and `Guid` - I have updated the method to call System.IsPrimitive() first and then check the other cases that are not .NET primitive but suitable for SQLite.**

 - Compute.ValidatePropertyMapping -> The name feels wrong as it implies that a Boolean is expected as return value. 
**Agreed, renamed to `Compute.PropertyMapping` which is cleaner.**

 - Compute.ResolveColumnSchema -> The separation in 3 tiers don't make sense to me and generate al lot of duplicate code. You can just collapse them into one: get all primitives that are not excluded, then add properties mapped by the user. With that in mind, I don't see the benefit of IRecord.
**Yeah the `IRecord` was just going to be make it an easy push, but given the mappings already pull out the primitive properties anyway, I agree.**

 - What's the point of the schema table if it is not used when pushing objects ?
**It's used when pulling objects back in the `ExecuteQuery` method. We could use it to store the `MappingInfo` but with `Fragments` it could get bloated.**

 - Query.GetPropertyValueFromPath() -> Why not use the GetPropertyValue method in the core engine ?
 **Agreed, removed.**

 - I'm not a big fan of the Example classes being defined in the oM. They should be in the Tests project instead
**Agreed, moved.** 

 - It is better to make your fallback method for ExecuteCommand private as it should never be called directly. So, making it private avoid the confusing with IExecuteCommand
**Agreed, updated.**

 - I don't see any way to remove entries from a table. I see two options: 
   - The user has an option (in the push config) to replace existing content with the one they are pushing
   - The user can explicitly delete the content of a table as an Execute command
I see you have a Compute.Delete method but I don't see how this fits into the current workflow

**I have implemented the `IDelete` override from the BHoMAdapter that takes an `IRequest` and `ActionConfig` - I've unified the `ExecuteQuery` to handle SELECT, DELETE and COUNT queries. So this is basically your first bullet where users can delete using an `IRequest`.** 

 - A bit strange that EqualityFilterRequest is using a list of ColumnFilter while RangeFilterRequest is using a list of dictionaries of GeneralDomains. Wouldn't it make more sense if they were structured the same way ?
**`ColumnFilter` I wanted to have one column and multiple values (e.g. `NodeId` with values 1,2,3,4,10,11,12 without having to repeat the `ColumnName` for each value. The `Dictionary<>` component in Grasshopper did not like me trying to use a List<object> as a value so I made the `ColumnFilter` object. The `Dictionary<string,GeneralDomain` I don't think requires another object because it would just be a wrapper for the dictionary.**

 - The methods EqualityFilterRequest and RangeFilterRequest are very similar (only one line different) so they should probably be handled directly in the Pull method. 
**Agreed, added `ISqlRequest` so they can share all the properties that are used and have a single statement for the conversion of each type.**

 - FilterResult is a strange name in this case. It is not the result of the FilterRequest as the name would implies. I am also not sure about the fact that is is created in a Convert method. Convert methods are usually used to convert BHoM object to and from external classes
 
 .
**Renamed to `FilterCommand` so it's consistent with the `TableCommand` above.**

adecler reviewed (commented):
My comments have generally been addressed. Just a few more:
 - The example folder in the `oM` still contains `SensorReading`
 -  In the Pull (line 63) you are checking for the type to decide which method to call. You should use our standard pattern of creating an `IFilterCommand(ISqlRequest)` method that uses a dynamic to dispatch to the appropriate method internally. 
 - Be careful that, although you are now supporting `IObject` instead of `IBHoMObject` you still have a few places where `IBHoMObject` is used instead of `IObject`
 -  The file for `FilterCommand` is still called `FilterResult`
 - You have multiple lines of `<Content Remove>` in your project file. Not sure what they are for ?

On an additional note, you are currently relying on the user to close the connection to the database. It would be good to also lose the connection when the adapter is deleted.

peterjamesnugent committed:
Remove examples from oM, follow through use of IObject - only exception is for Fragment extraction which requires an IBHoMObject

peterjamesnugent commented:
- The example folder in the `oM` still contains `SensorReading`
**Removed**

 -  In the Pull (line 63) you are checking for the type to decide which method to call. You should use our standard pattern of creating an `IFilterCommand(ISqlRequest)` method that uses a dynamic to dispatch to the appropriate method internally. 
**I should have done that in the first place, updated.**

 - Be careful that, although you are now supporting `IObject` instead of `IBHoMObject` you still have a few places where `IBHoMObject` is used instead of `IObject`
**Removed these, only exception where `IBHoMObject` is used is for `Fragment` extraction which a property of `IBHoMObject`.**

 -  The file for `FilterCommand` is still called `FilterResult`
**Updated**

 - You have multiple lines of `<Content Remove>` in your project file. Not sure what they are for ?
**It's the e_sqlite3.dll that shows up in SQL_Engine project and doesn't get removed. I have removed the content remove tags. I have also removed the references to SQL from the Engine as all those methods sit in the Adapter.**

On an additional note, you are currently relying on the user to close the connection to the database. It would be good to also lose the connection when the adapter is deleted.
**I have added a check for `m_active` to close if `active` from `SQLiteAdapter` is closed.**

All methods that generate commands for SQL are stored in `BH.Engine.SQLite.Compute` so they are accessible from outside the toolkit. The SQLite_Adapter/Helpers are just wrappers that call the `Execute` method using the `BH.oM.SQLiteCommand` generated by the `Compute` methods. We could just call the `Execute` method directly but the helper methods deal with the outputs being the correct `type`. I think this is much cleaner, allows access to the commands outside the adapter and keeps negates the need to reference `SQLite` in the `Engine`.

peterjamesnugent committed:
Use dynamic casting to deal with Pul, rename FilterCommand file and remove <Remove Content> commands

peterjamesnugent committed:
Migrate the functionality for computing SQL commands to the Compute class to allow exposure to external toolkits.  Each method is appended with a Command at the end - the Adapter simply calls these Compute methods and runs it through the Execute.

peterjamesnugent committed:
Move commands in to a commands folder and remove any redundant commands

adecler commented:
When activating the adapter, I get the following errors. Everything else works fine though. 

<img width="2090" height="579" alt="image" src="https://github.com/user-attachments/assets/db2cf6ac-3c95-4aa5-982f-7a67577004c2" />

adecler commented:
After the last commit, I don't get the error above anymore.
That being said, the database is not released correctly after turning off the adapters and closing the GH script (not Rhino):

<img width="927" height="502" alt="image" src="https://github.com/user-attachments/assets/29b09fbf-5cf7-494e-9723-dd936bd33eea" />

peterjamesnugent committed:
Add sorting for filter queries

peterjamesnugent committed:
Make m_connection static so that we are able to close the database when setting the adapter to false. and use garbage collection to remove temp files and relinquish access

peterjamesnugent committed:
Add WalCheckpointMode enum for use across the adapter

peterjamesnugent committed:
Add a check to prevent recursion of a WALCheckpoint command

peterjamesnugent committed:
Update CheckJournalMode and WalCheckpointCommand to not use Execute method to avoid recursion

peterjamesnugent committed:
Remove redundant null checks for adapter methods that are called from methods that already check null

peterjamesnugent commented:
@BHoMBot check required

peterjamesnugent committed:
Revert making m_connection static and implement destructor

adecler commented:
I still have the issue where the database file cannot be deleted unless I close Rhino completely. It might be worth triggering the garbage collector when `active` is turned to false. It doesn't solve everything but it's better than nothing.

peterjamesnugent committed:
Rename TableCommand

peterjamesnugent committed:
Add garbage collection for adapter to force destructor

adecler reviewed (approved):
My comments have been addressed. Happy to approve.

We will probably have to add additional support on the BHoM UI to make sure that the adapters are disposed of quickly but this is good enough for now.
